src/main/java/slimeknights/tconstruct/tables/inventory/table/PartBuilderContainer.java

```diff
@@ -148,4 +155,71 @@ public void onCrafting(PlayerEntity playerEntity, ItemStack output, IInventory c
   public boolean canMergeSlot(ItemStack stack, Slot slotIn) {
     return slotIn.inventory != this.craftResult && super.canMergeSlot(stack, slotIn);
   }
+
+  @OnlyIn(Dist.CLIENT)
+  public List<PartRecipe> getPartRecipeList() {
+    return this.partRecipes;
+  }
+
+  @OnlyIn(Dist.CLIENT)
+  public int getPartRecipeListSize() {
+    return this.partRecipes.size();
+  }
+
+  @OnlyIn(Dist.CLIENT)
+  public boolean hasPatternInPatternSlot() {
+    return this.patternSlot.getHasStack() && !this.partRecipes.isEmpty();
+  }
+
+  /**
+   * Returns the index of the selected recipe.
+   */
+  @OnlyIn(Dist.CLIENT)
+  public int getSelectedPartRecipe() {
+    return this.selectedRecipe.get();
+  }
+
+  /**
+   * Handles the given Button-click on the server, currently only used by enchanting. Name is for legacy.
+   */
+  @Override
+  public boolean enchantItem(PlayerEntity playerIn, int id) {
+    if (id >= 0 && id < this.partRecipes.size()) {
+      this.selectedRecipe.set(id);
+      this.updateResult();
+    }
+
+    return true;
+  }
+
+  private void updateAvailablePartRecipes(IInventory inventoryIn, ItemStack stack) {
+    this.partRecipes.clear();
+    this.selectedRecipe.set(-1);
+    this.craftResult.setInventorySlotContents(0, ItemStack.EMPTY);
+
+    if (!stack.isEmpty()) {
+      this.partRecipes = this.world.getRecipeManager().getRecipes(TinkerTables.partRecipeType, inventoryIn, this.world);
```

Comment:
Instead of each part recipe being a button, I would rather patterns be separate from part recipes. So basically, iterate all recipes to get a list of unique patterns. This will allow us to add non-tool part recipes later, such as an ingot pattern (mostly for modpacks).

We could work on this in a later commit instead of here.